### ISSUE TITLE
Out-of-Memory errors with heavy(?) writes.

### ISSUE THREAD

**Issue by ndouglas** [memcached/memcached#793]
As some background, we have [a Drupal codebase](https://github.com/department-of-veterans-affairs/va.gov-cms) that we run in three different configurations: locally, via [Lando](https://lando.dev/); PR/CI/Demo, via [Tugboat](https://www.tugboat.qa/), and "BRD": our dev/staging/prod app deploy.  We have several suites of tests (PHPUnit, Cypress, Behat, etc) that run against the codebase.  By default, these tests are run in Tugboat and on Staging post-deploy as a last-minute check before deploying to production.

Several weeks ago, we integrated Memcached into these three configurations, via the Bitnami Docker image for Lando, [Tugboat's slightly customized Docker image](https://github.com/TugboatQA/dockerfiles/blob/main/memcached/1.6.9/Dockerfile) for Tugboat, and Memcached via ElastiCache on BRD.  For the most part, this has gone pretty smoothly.  We're running 1.6.9 in both Lando and Tugboat, and so far as I can tell there have been no alterations to the source code, and the settings seem to be left at defaults in most or all cases:

```
STAT maxbytes 268435456
STAT maxconns 1024
STAT tcpport 11211
STAT udpport 0
STAT inter NULL
STAT verbosity 0
STAT oldest 0
STAT evictions on
STAT domain_socket NULL
STAT umask 700
STAT growth_factor 1.25
STAT chunk_size 48
STAT num_threads 4
STAT num_threads_per_udp 4
STAT stat_key_prefix :
STAT detail_enabled no
STAT reqs_per_event 20
STAT cas_enabled yes
STAT tcp_backlog 1024
STAT binding_protocol auto-negotiate
STAT auth_enabled_sasl no
STAT auth_enabled_ascii no
STAT item_size_max 1048576
STAT maxconns_fast yes
STAT hashpower_init 0
STAT slab_reassign yes
STAT slab_automove 1
STAT slab_automove_ratio 0.80
STAT slab_automove_window 30
STAT slab_chunk_max 524288
STAT lru_crawler yes
STAT lru_crawler_sleep 100
STAT lru_crawler_tocrawl 0
STAT tail_repair_time 0
STAT flush_enabled yes
STAT dump_enabled yes
STAT hash_algorithm murmur3
STAT lru_maintainer_thread yes
STAT lru_segmented yes
STAT hot_lru_pct 20
STAT warm_lru_pct 40
STAT hot_max_factor 0.20
STAT warm_max_factor 2.00
STAT temp_lru no
STAT temporary_ttl 61
STAT idle_timeout 0
STAT watcher_logbuf_size 262144
STAT worker_logbuf_size 65536
STAT read_buf_mem_limit 0
STAT track_sizes no
STAT inline_ascii_response no
STAT ext_item_size 512
STAT ext_item_age 4294967295
STAT ext_low_ttl 0
STAT ext_recache_rate 2000
STAT ext_wbuf_size 4194304
STAT ext_compact_under 0
STAT ext_drop_under 0
STAT ext_max_frag 0.80
STAT slab_automove_freeratio 0.010
STAT ext_drop_unread no
STAT ssl_enabled no
STAT ssl_chain_cert (null)
STAT ssl_key (null)
STAT ssl_verify_mode 0
STAT ssl_keyformat 1
STAT ssl_ciphers NULL
STAT ssl_ca_cert NULL
STAT ssl_wbuf_size 16384
STAT ssl_session_cache no
STAT num_napi_ids (null)
STAT memory_file (null)
END
```

A couple of weeks ago, however, we started seeing failing CI tests due to OOM errors in Memcached on Tugboat.  [Here](https://github.com/department-of-veterans-affairs/va.gov-cms/pull/5386#issuecomment-844294967) is a link to the PR where I first saw the failure.  Subsequent commits passed tests and I haven't seen any reoccurrences of this specific failure, so I assumed it was a blip and moved on with my life.

A few days after that -- maybe two weeks ago -- we started seeing OOM errors more consistently, on one specific test.  Without getting too much into detail, this particular test seems to me to place some significant demands on Memcached.  It loads a very large data structure into memory, builds a table from it, and then compares that table to an expected table.  So I can understand why it would stress Memcache, but OOM errors sound like a (relatively) serious failure to me as opposed to some suboptimal performance, hiccup, etc.

By creating a lot of PRs at the same time, we were able to reproduce this some substantial percentage of the time.  Some other issues confounded the problem, so I shifted to trying to trigger the problem directly -- no Drupal, no Memcache[d] PHP modules, just `memaslap` and then `mc-crusher`.  

With `memaslap`, I ran `./clients/memaslap -s memcache:11211 -t 360s` on both an instance that had been built prior to when we saw these issues arise and instance that had been built after we started seeing those issues.  The former seemed to complete without any issues; the latter displayed, among the normal lines of output, about a thousand repetitions of `SERVER_ERROR out of memory storing object`.  (Further output is in comments [here](https://github.com/department-of-veterans-affairs/va.gov-cms/issues/5491#issuecomment-854182054) and [here](https://github.com/department-of-veterans-affairs/va.gov-cms/issues/5491#issuecomment-854226586).

With `mc-crusher`, I found that I could reliably create OOM errors -- `mc-crusher would not report them`, but they'd appear in the output of `memcache-tool-ng` -- just by bouncing between the conf files once Memcached was utilizing all of its allocated memory.

For example, this is the output of `memcache-tool-ng` from an instance that has failed a CI test:

```
memcache:11211:
   #  Item_Size    Max_age   Pages     Count   Used Mem     To Store   Efficiency              Evicted  Evict_Time          OOM
   2       120B       889s       1        14    1.00 MB      1.51 KB       0.15 %                    0           0            0
   3       152B       889s       1         4    1.00 MB        494 b       0.05 %                    0           0            0
   4       192B         0s       1         0    1.00 MB          0 b       0.00 %                    0           0            0
   5       240B         0s       1         0    1.00 MB          0 b       0.00 %                    0           0            0
   6       304B     39555s       1        53    1.00 MB     14.91 KB       1.46 %                    0           0            0
   7       384B     40276s       2      3333    2.00 MB      1.14 MB      56.75 %                    0           0            0
   8       480B     40276s       2      3693    2.00 MB      1.42 MB      70.89 %                    0           0            0
   9       600B     39599s       1       285    1.00 MB    155.09 KB      15.15 %                    0           0            0
  10       752B     39590s       1       449    1.00 MB    286.70 KB      28.00 %                    0           0            0
  11       944B     39678s       1       856    1.00 MB    705.57 KB      68.90 %                    0           0            0
  12       1.2K     38837s       2      1138    2.00 MB      1.18 MB      59.08 %                  359         369            0
  13       1.4K     38751s       3      1419    3.00 MB      1.92 MB      64.09 %                 2315       36622            1
  14       1.8K     38752s       4      1831    4.00 MB      2.95 MB      73.77 %                 1144         605            0
  15       2.3K     39293s      16      7215   16.00 MB     13.89 MB      86.80 %                 2410       38133            0
  16       2.8K     39308s      18      6149   18.00 MB     16.31 MB      90.61 %                 5849       37307            0
  17       3.5K     39303s      92     26496   92.00 MB     81.57 MB      88.66 %                25729       38417            0
  18       4.4K     39260s      41      9430   41.00 MB     36.71 MB      89.54 %                 6389       38374            0
  19       5.5K     39302s      25      4600   25.00 MB     21.71 MB      86.84 %                 2136       38403            0
  20       6.9K     39308s      10      1458   10.00 MB      8.75 MB      87.52 %                  756       37209            0
  21       8.7K     38822s       3       243    3.00 MB      1.78 MB      59.30 %                  167       36791            0
  22      10.8K     39291s       2       188    2.00 MB      1.77 MB      88.67 %                  103       38393            0
  23      13.6K     39554s       2       150    2.00 MB      1.76 MB      87.84 %                    2        1475            0
  24      16.9K     38834s       2       120    2.00 MB      1.82 MB      90.78 %                  161         813            0
  25      21.2K     38667s       3       143    3.00 MB      2.50 MB      83.37 %                  189       36366            0
  26      26.5K     38227s       1        37    1.00 MB    901.09 KB      88.00 %                  149         245            0
  27      33.1K     38546s       1        30    1.00 MB    909.89 KB      88.86 %                   26       33756            0
  28      41.4K     39307s       1         6    1.00 MB    228.62 KB      22.33 %                    0           0            0
  29      51.7K     39306s       1        13    1.00 MB    608.14 KB      59.39 %                    0           0            0
  30      64.7K     39304s       2        21    2.00 MB      1.23 MB      61.28 %                    0           0            0
  31      80.9K     39304s       2        23    2.00 MB      1.64 MB      81.96 %                   26         484            0
  32     101.1K     39310s       2        13    2.00 MB      1.15 MB      57.45 %                    0           0            0
  33     126.3K     39294s       3        17    3.00 MB      1.85 MB      61.60 %                    0           0            0
  34     157.9K     39280s       1         5    1.00 MB    693.78 KB      67.75 %                    6         453            0
  35     197.4K     39280s       2         7    2.00 MB      1.18 MB      58.91 %                    0           0            0
  36     246.8K     39293s       1         4    1.00 MB    839.47 KB      81.98 %                    4         964            0
  37     308.5K     39287s       1         2    1.00 MB    509.40 KB      49.75 %                    0           0            0
  38     385.6K      7286s       3         6    3.00 MB      1.93 MB      64.37 %                   72       33278            0

Total used memory: 256.00 MB (for 211.86 MB of data)
Global efficiency:  82.76 %
```

(256MB allocated)

If I run `mc-crusher` with `slab_rebal_torture` for, say, 30 seconds, I see the following:

```
memcache:11211:
   #  Item_Size    Max_age   Pages     Count   Used Mem     To Store   Efficiency              Evicted  Evict_Time          OOM
   1        96B         6s       3     32766    3.00 MB      2.21 MB      73.61 %                62446           3          207
   2       120B         6s       1      8738    1.00 MB    853.32 KB      83.33 %                88953           3            0
   3       152B         6s       1      6898    1.00 MB    943.09 KB      92.10 %                80151           3            0
   4       192B        11s       4     21844    4.00 MB      3.53 MB      88.27 %                78009           8            0
   5       240B        11s       6     26213    6.00 MB      5.49 MB      91.49 %                72804           8            0
   6       304B         5s       1      3449    1.00 MB    909.40 KB      88.81 %                94243           2            0
   7       384B         5s       2      5460    2.00 MB      1.92 MB      96.12 %                91263           2            0
   8       480B         5s       2      4368    2.00 MB      1.96 MB      97.78 %               100517           2            0
   9       600B         5s       2      3494    2.00 MB      1.90 MB      94.97 %                91109           2            0
  10       752B         4s       1      1394    1.00 MB    911.04 KB      88.97 %                90519           1            0
  11       944B         5s       4      4440    4.00 MB      3.68 MB      92.07 %                88317           2            0
  12       1.2K         5s       5      3653    5.00 MB      3.73 MB      74.54 %                95991           2            0
  13       1.4K         5s       4      2832    4.00 MB      3.70 MB      92.50 %                91403           2            1
  14       1.8K         4s       5      2255    5.00 MB      3.81 MB      76.11 %                48619           1            0
  15       2.3K     39497s      16      7215   16.00 MB     13.89 MB      86.80 %                 2410       38133            0
  16       2.8K     39020s       2       722    2.00 MB      1.89 MB      94.38 %                 5851       39504            0
  17       3.5K     39507s      92     26495   92.00 MB     81.57 MB      88.66 %                25730       39479            0
  18       4.4K     39464s      41      9430   41.00 MB     36.71 MB      89.54 %                 6389       38374            0
  19       5.5K     39506s      25      4599   25.00 MB     21.71 MB      86.82 %                 2137       39478            0
  20       6.9K     39512s       7      1029    7.00 MB      6.20 MB      88.61 %                  756       37209            0
  21       8.7K     39026s       3       243    3.00 MB      1.78 MB      59.30 %                  167       36791            0
  22      10.8K     39491s       2       186    2.00 MB      1.76 MB      87.82 %                  104       39467            0
  23      13.6K     39758s       2       149    2.00 MB      1.74 MB      87.25 %                    3       39730            0
  24      16.9K     39038s       2       120    2.00 MB      1.82 MB      90.78 %                  161         813            0
  25      21.2K     38871s       3       143    3.00 MB      2.50 MB      83.37 %                  189       36366            0
  26      26.5K     38431s       1        37    1.00 MB    901.09 KB      88.00 %                  149         245            0
  27      33.1K     38750s       1        30    1.00 MB    909.89 KB      88.86 %                   26       33756            0
  28      41.4K     39511s       1         6    1.00 MB    228.62 KB      22.33 %                    0           0            0
  29      51.7K     39510s       1        13    1.00 MB    608.14 KB      59.39 %                    0           0            0
  30      64.7K     39508s       2        21    2.00 MB      1.23 MB      61.28 %                    0           0            0
  31      80.9K     39508s       2        23    2.00 MB      1.64 MB      81.96 %                   26         484            0
  32     101.1K     39514s       2        13    2.00 MB      1.15 MB      57.45 %                    0           0            0
  33     126.3K     39498s       3        17    3.00 MB      1.85 MB      61.60 %                    0           0            0
  34     157.9K     39484s       1         5    1.00 MB    693.78 KB      67.75 %                    6         453            0
  35     197.4K     39484s       2         7    2.00 MB      1.18 MB      58.91 %                    0           0            0
  36     246.8K     39497s       1         4    1.00 MB    839.47 KB      81.98 %                    4         964            0
  37     308.5K     39491s       1         2    1.00 MB    509.40 KB      49.75 %                    0           0            0
  38     385.6K      7486s       3         5    3.00 MB      1.61 MB      53.65 %                   73        7462            0

Total used memory: 257.00 MB (for 220.24 MB of data)
Global efficiency:  85.70 %
```

with `slab_rebal_torture2`, we see:

```
memcache:11211:
   #  Item_Size    Max_age   Pages     Count   Used Mem     To Store   Efficiency              Evicted  Evict_Time          OOM
   1        96B         6s       3     32766    3.00 MB      2.21 MB      73.71 %               143256           4          651
   2       120B         5s       1      8738    1.00 MB    853.32 KB      83.33 %               236334           3            0
   3       152B         4s       1      6898    1.00 MB    943.09 KB      92.10 %               227118           2            0
   4       192B         5s       4     21844    4.00 MB      3.53 MB      88.29 %               177639           3          340
   5       240B         5s       6     26213    6.00 MB      5.49 MB      91.55 %               135016           3          355
   6       304B         3s       3      4042    3.00 MB      1.04 MB      34.56 %               240268           1            0
   7       384B         4s       2      5460    2.00 MB      1.93 MB      96.33 %               234562           2            0
   8       480B         3s       2      4368    2.00 MB      1.96 MB      97.89 %               246618           1            0
   9       600B         3s       3      5240    3.00 MB      2.84 MB      94.79 %               230209           1           46
  10       752B        13s      11     15055   11.00 MB      9.62 MB      87.50 %               161244           0           73
  11       944B         3s      10      9990   10.00 MB      8.29 MB      82.88 %               180096           1          208
  12       1.2K         3s       8      7080    8.00 MB      7.22 MB      90.29 %               166465           1          403
  13       1.4K         3s       4      2832    4.00 MB      3.70 MB      92.48 %               181768           1          301
  14       1.8K        14s       5      2502    5.00 MB      4.22 MB      84.44 %                48619           1            0
  15       2.3K     39562s      16      7215   16.00 MB     13.89 MB      86.80 %                 2410       38133            0
  16       2.8K     39084s       2       721    2.00 MB      1.89 MB      94.26 %                 5854       39051            0
  17       3.5K     39572s      75     21600   75.00 MB     66.44 MB      88.59 %                25731       39559            0
  18       4.4K     39529s      41      9430   41.00 MB     36.71 MB      89.54 %                 6389       38374            0
  19       5.5K     39571s      25      4599   25.00 MB     21.71 MB      86.82 %                 2137       39478            0
  20       6.9K     39549s       2       294    2.00 MB      1.76 MB      87.94 %                  784       39537            0
  21       8.7K     39091s       3       243    3.00 MB      1.78 MB      59.30 %                  167       36791            0
  22      10.8K     39556s       2       186    2.00 MB      1.76 MB      87.82 %                  104       39467            0
  23      13.6K     39823s       2       149    2.00 MB      1.74 MB      87.25 %                    3       39730            0
  24      16.9K     39103s       2       120    2.00 MB      1.82 MB      90.78 %                  161         813            0
  25      21.2K     38936s       3       143    3.00 MB      2.50 MB      83.37 %                  189       36366            0
  26      26.5K     38496s       1        37    1.00 MB    901.09 KB      88.00 %                  149         245            0
  27      33.1K     38815s       1        30    1.00 MB    909.89 KB      88.86 %                   26       33756            0
  28      41.4K     39576s       1         6    1.00 MB    228.62 KB      22.33 %                    0           0            0
  29      51.7K     39575s       1        13    1.00 MB    608.14 KB      59.39 %                    0           0            0
  30      64.7K     39573s       2        21    2.00 MB      1.23 MB      61.28 %                    0           0            0
  31      80.9K     39573s       2        23    2.00 MB      1.64 MB      81.96 %                   26         484            0
  32     101.1K     39579s       2        13    2.00 MB      1.15 MB      57.45 %                    0           0            0
  33     126.3K     39563s       3        17    3.00 MB      1.85 MB      61.60 %                    0           0            0
  34     157.9K     39549s       1         5    1.00 MB    693.78 KB      67.75 %                    6         453            0
  35     197.4K     39549s       2         7    2.00 MB      1.18 MB      58.91 %                    0           0            0
  36     246.8K     39562s       1         4    1.00 MB    839.47 KB      81.98 %                    4         964            0
  37     308.5K     39556s       1         2    1.00 MB    509.40 KB      49.75 %                    0           0            0
  38     385.6K      7551s       3         5    3.00 MB      1.61 MB      53.65 %                   73        7462            0

Total used memory: 257.00 MB (for 219.03 MB of data)
Global efficiency:  85.23 %
```

I see this to a lesser degree locally, in Lando, with a memory allocation of 64MB.  

```
memcache:11211:
   #  Item_Size    Max_age   Pages     Count   Used Mem     To Store   Efficiency              Evicted  Evict_Time          OOM
   1        96B        22s       1     10922    1.00 MB    755.99 KB      73.83 %              2614989          22           85
   2       120B        19s       1      8738    1.00 MB    852.77 KB      83.28 %              2837742          18            3
   3       152B        17s       1      6898    1.00 MB    943.09 KB      92.10 %              2724275          16            2
   4       192B        20s       2     10922    2.00 MB      1.77 MB      88.54 %              2601342          19            9
   5       240B        22s       2      8738    2.00 MB      1.83 MB      91.67 %              2662143          21            5
   6       304B        15s       2      6898    2.00 MB      1.77 MB      88.61 %              2820779          14            6
   7       384B        13s       2      5460    2.00 MB      1.92 MB      96.17 %              2826932          12            9
   8       480B        14s       2      4368    2.00 MB      1.96 MB      97.89 %              2605269          14            6
   9       600B        13s       2      3494    2.00 MB      1.90 MB      94.97 %              2648975          13            8
  10       752B        10s       2      2788    2.00 MB      1.78 MB      89.19 %              2762617          10            8
  11       944B         9s       2      2220    2.00 MB      1.84 MB      92.12 %              2589044           9           36
  12       1.2K         6s       2      1770    2.00 MB      1.81 MB      90.32 %              2597391           6           50
  13       1.4K         4s       2      1416    2.00 MB      1.85 MB      92.50 %              2364764           4          180
  14       1.8K         1s       7      3945    7.00 MB      6.66 MB      95.12 %              1203455           0         1844
  15       2.3K    510085s       2       901    2.00 MB      1.71 MB      85.27 %                14859      435811            0
  16       2.8K    510038s       2       721    2.00 MB      1.96 MB      97.89 %                10101      435823            0
  17       3.5K    507574s       2       575    2.00 MB      1.79 MB      89.59 %                61694      433359            0
  18       4.4K    507572s       2       460    2.00 MB      1.80 MB      89.87 %                42058      170495            0
  19       5.5K    507573s       2       367    2.00 MB      1.72 MB      85.92 %                17653      433357            0
  20       6.9K    507574s       2       293    2.00 MB      1.76 MB      87.77 %                 4268      433300            0
  21       8.7K    510924s       2       235    2.00 MB      1.75 MB      87.39 %                  457      436651            0
  22      10.8K   1121597s       2       184    2.00 MB      1.73 MB      86.30 %                   28      608240            0
  23      13.6K   1121601s       2       149    2.00 MB      1.76 MB      87.79 %                    4      608244            0
  24      16.9K   1121598s       2       120    2.00 MB      1.78 MB      88.81 %                   23      608244            0
  25      21.2K   1186412s       2        78    2.00 MB      1.42 MB      71.09 %                    0           0            0
  26      26.5K   1121606s       1        25    1.00 MB    599.08 KB      58.50 %                    0           0            0
  27      33.1K    599811s       1        16    1.00 MB    485.87 KB      47.45 %                    0           0            0
  28      41.4K    510931s       1         5    1.00 MB    185.35 KB      18.10 %                    0           0            0
  29      51.7K    511414s       1        12    1.00 MB    557.58 KB      54.45 %                    0           0            0
  30      64.7K    510832s       1        15    1.00 MB    891.92 KB      87.10 %                   26        3291            0
  31      80.9K    510730s       1        12    1.00 MB    900.54 KB      87.94 %                   33         705            0
  32     101.1K    510483s       1        10    1.00 MB    924.53 KB      90.29 %                   20         479            0
  33     126.3K    510357s       1         8    1.00 MB    910.30 KB      88.90 %                   24         358            0
  34     157.9K    510377s       1         6    1.00 MB    847.21 KB      82.73 %                    7        2854            0
  35     197.4K    510315s       1         5    1.00 MB    842.42 KB      82.27 %                    6        1487            0
  36     246.8K    510582s       1         4    1.00 MB    829.30 KB      80.99 %                    5         371            0
  37     308.5K    510604s       1         1    1.00 MB    252.56 KB      24.66 %                    0           0            0
  38     385.6K   1021060s       2         4    2.00 MB      1.29 MB      64.38 %                    0           0            0

Total used memory: 66.00 MB (for 56.07 MB of data)
Global efficiency:  84.95 %
```

The conditions are harder to reproduce on dev/staging/prod, with 13GB allocated, but after a couple hours of playing around we managed to fill up the cache, get a decent number of evictions, but no OOM errors:

```
10.247.97.110:11211:
   #  Item_Size    Max_age   Pages     Count   Used Mem     To Store   Efficiency              Evicted  Evict_Time          OOM
   1        96B       642s     240   2621280  240.00 MB    181.96 MB      75.82 %               874686         411            0
   2       120B       671s     295   2577710  295.00 MB    250.13 MB      84.79 %               690483         440            0
   3       152B       693s     348   2400504  348.00 MB    324.39 MB      93.22 %               563389         462            0
   4       192B       702s     463   2528443  463.00 MB    414.05 MB      89.43 %               560411         471            0
   5       240B       707s     518   2263142  518.00 MB    478.40 MB      92.35 %               477498         475            0
   6       304B       710s     651   2245299  651.00 MB    581.67 MB      89.35 %               463555         478            0
   7       384B      1819s     913   2492484  913.00 MB    883.24 MB      96.74 %               291734        1575            0
   8       480B       718s    1133   2474472    1.11 GB      1.09 GB      98.24 %               485240         486            0
   9       600B       721s    1170   2043990    1.14 GB      1.09 GB      95.23 %               401520         703            0
  10       752B       720s    1255   1749470    1.23 GB      1.10 GB      89.40 %               350203         705            0
  11       944B       724s    1221   1355310    1.19 GB      1.10 GB      92.34 %               268548         709            0
  12       1.2K       333s    2742   2426670    2.68 GB      2.24 GB      83.75 %              1672346         333            0
  13       1.4K      2211s    1413   1000230    1.38 GB      1.28 GB      92.61 %                    0           0            0
  14       1.8K      2211s     887    500144  887.00 MB    845.08 MB      95.27 %                    0           0            0
  15       2.3K      2210s       1        66    1.00 MB    133.45 KB      13.03 %                    0           0            0
  16       2.8K      2209s       1        71    1.00 MB    188.06 KB      18.37 %                    0           0            0
  17       3.5K      2210s       8      2202    8.00 MB      6.77 MB      84.62 %                    0           0            0
  18       4.4K      2210s       7      1492    7.00 MB      5.88 MB      84.00 %                    0           0            0
  19       5.5K      2209s       3       414    3.00 MB      2.01 MB      67.05 %                    0           0            0
  20       6.9K      2210s       3       330    3.00 MB      1.97 MB      65.55 %                    0           0            0
  21       8.7K      2210s       1        78    1.00 MB    600.25 KB      58.62 %                    0           0            0
  22      10.8K      2204s       1        24    1.00 MB    233.94 KB      22.85 %                    0           0            0
  23      13.6K      2204s       1        19    1.00 MB    227.50 KB      22.22 %                    0           0            0
  24      16.9K      2210s       1        23    1.00 MB    349.97 KB      34.18 %                    0           0            0
  25      21.2K      2204s       1        16    1.00 MB    295.55 KB      28.86 %                    0           0            0
  26      26.5K      1010s       1         1    1.00 MB     22.76 KB       2.22 %                    0           0            0
  27      33.1K      1031s       1         3    1.00 MB     87.61 KB       8.56 %                    0           0            0
  28      41.4K         0s       1         0    1.00 MB          0 b       0.00 %                    0           0            0
  29      51.7K      1011s       1         2    1.00 MB    100.06 KB       9.77 %                    0           0            0
  30      64.7K         0s       1         0    1.00 MB          0 b       0.00 %                    0           0            0
  31      80.9K       938s       1         1    1.00 MB     76.65 KB       7.49 %                    0           0            0
  32     101.1K        14s       1         1    1.00 MB     89.29 KB       8.72 %                    0           0            0
  35     197.4K      2204s       1         2    1.00 MB    372.65 KB      36.39 %                    0           0            0
  37     308.5K      1010s       1         1    1.00 MB    255.30 KB      24.93 %                    0           0            0
  38     385.6K       939s       1         1    1.00 MB    329.63 KB      32.19 %                    0           0            0

Total used memory: 12.98 GB (for 11.78 GB of data)
Global efficiency:  90.77 %
```

So after working on this for a while, off and on, I feel like I know a bit more about Memcached but not much more about this specific problem.  Ultimately, I guess I have a few questions that you might be able to answer quickly:

1.  Are these OOM errors to be expected once we have a full cache and continue to attempt to cache new items?  Is this something the application should be expected to handle?

2.  Does anything jump out at you about our configuration, general structure, or anything else as potentially problematic?

If there is any other information I can provide, I'll be happy to pass it along.  We're doing our best to diagnose this issue from our end, and would deeply appreciate any guidance or suggestions or other feedback you can provide 😅

**Comment by dormando** [comment]
Hey,

I'll start off by saying it looks like you're using elasticache (though sounds like they shipped 1.6.9?) so support will be limited; they only randomly update the binary and may have custom patches that change behavior.

To summarize, yes that's an edge case. Especially with low memory caps (256-64 megs) and large objects (> 256k), the instance will not have very much memory available to handle object uploads. Data is uploaded directly into the final object memory, so if slab class 38 only has 1 megabyte of memory in it, the instance can only handle 1 object upload at a time. On a larger instance (a gigabyte+) there's a pool of free memory, and the memory balancer which should move extra data pages into the higher classes over time... but it's not useful for sudden changes.

This is something we should improve in the future but I don't have any present plans... I can give you a few things to try (especially for your small memory/temporary PR testing thing?)

`-o slab_chunk_max=8192` (or even 4096 or so). This uses chunk stitching to create larger objects, and setting the max lower will create fewer slab classes, improving the odds of there being pages where you do want to store something. It's not well tested to set this so low (default is 512k) on a large production host but in testing on small memory it should be okay.

Similarly you can adjust the class growth factor (`-f`) to create fewer slab classes (test with `memcached -vvv` to get a quick dump.

Let me know if those help at all?

**Comment by ndouglas** [comment]
Thanks for your (swift!) reply!

Sorry, I was confusing -- yeah, we're using ElastiCache but it's not displaying this issue -- elsewhere we use Memcached directly.

After banging our heads against it for a few more hours today, I think we've reached the same conclusion you just gave: that we just set the memory limits too low for the peak workload we're trying to send its way.  Some of our tests create quite large data structures, and I suspect that they're overwhelming a particular slab just as you describe.

I deeply appreciate your response!  We'll work with your suggestions and we'll almost certainly be upping the memory allocation, because it sounds like we're really underprovisioning.

Thanks again!

**Comment by dormando** [comment]
You may still have issues if the memory usage is very sudden.. the memory balancer has gotten a lot faster over the years but it's still generally tuned for large instances that changes its memory profile over the course of a day rather than in seconds.

Setting chunk max down to 256k or 128k will probably be fine for larger instances anyway; I do want to reduce that value's default over time.

The TODO pile is simply too large, but in the future we could delay or queue connections instead of immediately OOM'ing. Not easy to do at the moment.

**Comment by sergmour** [comment]
@dormando : Hey Dormando, we also hit a similar issue very consistently with ExtStore feature. In short, our customers expect items to be moved from RAM into SSD when RAM limit is hit. And this does happen, but it looks like the metadata is moved to lower slabs and these slabs get to a point when there is no more space to grow ending up with slab-reassign-evictions. We do set oslab_automove_freeratio between 10-25% and it helps a lot, but it does not solve the problem completely especially for long-running batch ingestion jobs.
So, in short, we'd like to get rid of these evictions but looks like we can't. Any thoughts/suggestions would be greatly appreciated.

**Comment by dormando** [comment]
@sergmour can you open a new issue with some details? It may look related but it's kind of not.

**Comment by sergmour** [comment]
done: https://github.com/memcached/memcached/issues/794 . Thanks!
